watertap/tools/recursive_parameter_sweep.py

```diff
@@ -0,0 +1,293 @@
+###############################################################################
+# WaterTAP Copyright (c) 2021, The Regents of the University of California,
+# through Lawrence Berkeley National Laboratory, Oak Ridge National
+# Laboratory, National Renewable Energy Laboratory, and National Energy
+# Technology Laboratory (subject to receipt of any required approvals from
+# the U.S. Dept. of Energy). All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. These files are also available online at the URL
+# "https://github.com/watertap-org/watertap/"
+#
+###############################################################################
+import numpy as np
+import pyomo.environ as pyo
+import sys
+import os
+import itertools
+import warnings
+import copy, pprint
+import h5py
+
+from scipy.interpolate import griddata
+from enum import Enum, auto
+from abc import abstractmethod, ABC
+from idaes.core.util import get_solver
+
+from idaes.surrogate.pysmo import sampling
+from idaes.core.util.model_statistics import (
+    variables_in_activated_equalities_set,
+    expressions_set,
+    total_objectives_set,
+)
+from pyomo.core.base.block import TraversalStrategy
+from watertap.tools.parameter_sweep import (
+    _aggregate_results_arr,
+    _build_combinations,
+    _create_local_output_skeleton,
+    _create_global_output,
+    _default_optimize,
+    _divide_combinations,
+    _do_param_sweep,
+    _init_mpi,
+    _process_sweep_params,
+    _process_results_filename,
+    _save_results,
+    _write_outputs,
+)
+
+np.set_printoptions(linewidth=200)
+
+# ================================================================
+
+
+def _filter_recursive_solves(model, sweep_params, outputs, recursive_local_dict, comm):
+
+    # Figure out how many filtered solves did this rank actually do
+    filter_counter = 0
+    for case, content in recursive_local_dict.items():
+        filter_counter += sum(
+            content["solve_successful"]
+        )  # content["solve_successful"].count(filter_keyword)
+
+    # Now that we have all of the local output dictionaries, we need to construct
+    # a consolidated dictionary of successful solves.
+    local_filtered_dict = _create_local_output_skeleton(
+        model, sweep_params, outputs, filter_counter
+    )
+    local_filtered_dict["solve_successful"] = []
+
+    # Populate local_successful_outputs
+    offset = 0
+    for case_number, content in recursive_local_dict.items():
+        # Filter all of the sucessful solves
+        optimal_indices = list(
+            itertools.compress(
+                range(len(content["solve_successful"])), content["solve_successful"]
+            )
+        )  # [idx for idx, status in enumerate(content["solve_status"]) if status == "optimal"]
```

Comment:
```suggestion
        optimal_indices = [idx for idx, success in enumerate(content["solve_successful"]) if success]
```
More readable, IMHO.